File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ConstructRandomStruct.java
package io.github.brenbar.uapi.internal;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import io.github.brenbar.uapi.RandomGenerator;

public class ConstructRandomStruct {
    static Map<String, Object> constructRandomStruct(
            Map<String, UFieldDeclaration> referenceStruct, Map<String, Object> startingStruct,
            boolean includeOptionalFields, boolean randomizeOptionalFields, List<UTypeDeclaration> typeParameters,
            RandomGenerator randomGenerator) {

        final var sortedReferenceStruct = new ArrayList<>(referenceStruct.entrySet());
        Collections.sort(sortedReferenceStruct, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

        final var obj = new TreeMap<String, Object>();
        for (final var field : sortedReferenceStruct) {
            final var fieldName = field.getKey();
            final var fieldDeclaration = field.getValue();
            final var blueprintValue = startingStruct.get(fieldName);
            final var useBlueprintValue = startingStruct.containsKey(fieldName);
            final UTypeDeclaration typeDeclaration = fieldDeclaration.typeDeclaration;

            final Object value;
            if (useBlueprintValue) {
                value = typeDeclaration.generateRandomValue(blueprintValue, useBlueprintValue,
                        includeOptionalFields, randomizeOptionalFields, typeParameters, randomGenerator);
            } else {
                if (!fieldDeclaration.optional) {
                    value = typeDeclaration.generateRandomValue(null, false,
                            includeOptionalFields, randomizeOptionalFields, typeParameters, randomGenerator);
                } else {
                    if (!includeOptionalFields || (randomizeOptionalFields && randomGenerator.nextBoolean())) {
                        continue;
                    }
                    value = typeDeclaration.generateRandomValue(null, false,
                            includeOptionalFields, randomizeOptionalFields, typeParameters, randomGenerator);
                }
            }

            obj.put(fieldName, value);
        }
        return obj;
    }
}

File: lib/ts/src/internal/generation/ConstructRandomStruct.ts
import { RandomGenerator } from '../../RandomGenerator';
import { UFieldDeclaration } from '../../internal/types/UFieldDeclaration';

export function constructRandomStruct(
    referenceStruct: Record<string, UFieldDeclaration>,
    startingStruct: Record<string, any>,
    includeOptionalFields: boolean,
    randomizeOptionalFields: boolean,
    randomGenerator: RandomGenerator,
): Record<string, any> {
    const sortedReferenceStruct = Array.from(Object.entries(referenceStruct)).sort((e1, e2) => {
        const a = e1[0];
        const b = e2[0];
        for (let i = 0; i < Math.min(a.length, b.length); i++) {
            const charCodeA = a.charCodeAt(i);
            const charCodeB = b.charCodeAt(i);
            if (charCodeA !== charCodeB) {
                // If the characters are different, return the comparison result
                // where lowercase letters are considered greater than uppercase letters
                return charCodeA - charCodeB;
            }
        }
        // If one string is a prefix of the other, the shorter string comes first
        return a.length - b.length;
    });

    const obj: Record<string, any> = {};
    for (const [fieldName, fieldDeclaration] of sortedReferenceStruct) {
        const blueprintValue = startingStruct[fieldName];
        const useBlueprintValue = fieldName in startingStruct;
        const typeDeclaration = fieldDeclaration.typeDeclaration;

        let value: any;
        if (useBlueprintValue) {
            value = typeDeclaration.generateRandomValue(
                blueprintValue,
                useBlueprintValue,
                includeOptionalFields,
                randomizeOptionalFields,
                randomGenerator,
            );
        } else {
            if (!fieldDeclaration.optional) {
                value = typeDeclaration.generateRandomValue(
                    null,
                    false,
                    includeOptionalFields,
                    randomizeOptionalFields,
                    randomGenerator,
                );
            } else {
                if (!includeOptionalFields || (randomizeOptionalFields && randomGenerator.nextBoolean())) {
                    continue;
                }
                value = typeDeclaration.generateRandomValue(
                    null,
                    false,
                    includeOptionalFields,
                    randomizeOptionalFields,
                    randomGenerator,
                );
            }
        }

        obj[fieldName] = value;
    }

    return obj;
}
